pkg/controllers/mpi_job_controller.go

```diff
@@ -0,0 +1,997 @@
+// attempt to process it, by calling the syncHandler.
+func (c *MPIJobController) processNextWorkItem() bool {
+	obj, shutdown := c.queue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.queue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the work queue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the work queue and attempted again after a back-off
+		// period.
+		defer c.queue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the work queue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// work queue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// work queue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the work queue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.queue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// MPIJob resource to be synced.
+		if err := c.syncHandler(key); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.queue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the MPIJob resource
+// with the current status of the resource.
+func (c *MPIJobController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the MPIJob with this namespace/name.
+	mpiJob, err := c.mpiJobLister.MPIJobs(namespace).Get(name)
+	// The MPIJob may no longer exist, in which case we stop processing.
+	if errors.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("mpi job '%s' in work queue no longer exists", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Get the launcher Job for this MPIJob.
+	launcher, err := c.getLauncherJob(mpiJob)
+	if err != nil {
+		return err
+	}
+	// We're done if the launcher either succeeded or failed.
+	done := launcher != nil && (launcher.Status.Succeeded == 1 || launcher.Status.Failed == 1)
```

Comment:
What will happen if launcher is running?

Reply:
If the launcher is already running, the rest of the method is mostly noop.